code/game/objects/effects/anomalies/anomalies_ectoplasm.dm

```diff
@@ -0,0 +1,208 @@
+
+			if(istype(impacted_thing, /obj/structure/window))
+				var/obj/structure/window/window_to_damage = impacted_thing
+				window_to_damage.take_damage(rand(60, 90))
+				if(window_to_damage?.fulltile)
+					new /obj/effect/temp_visual/revenant/cracks(get_turf(window_to_damage))
+
+	if(effect_power >= 35)
+		var/effect_range = ghosts_orbiting + 3
+		haunt_outburst(epicenter = get_turf(src), range = effect_range, haunt_chance = 45, duration = 2 MINUTES)
+
+	if(effect_power >= 50) //Summon a ghost swarm!
+		var/list/candidate_list = list()
+		for(var/mob/dead/observer/orbiter in orbiters?.orbiter_list)
+			candidate_list += orbiter
+
+		new /obj/structure/ghost_portal(get_turf(src), candidate_list)
+
+		priority_announce("Anomaly has reached critical mass. Ectoplasmic outburst detected.", "Anomaly Alert")
+
+/**
+ * Manages updating the sprite for the anomaly based on how many orbiters it has.
+ *
+ *
+ * A check that is run to determine which sprite the anoamly should currently be displaying.
+ * With 50% or more participation, the "heavy" sprite is used. Otherwise, it is reverted to the normal anomaly sprite.
+ */
+
+/obj/effect/anomaly/ectoplasm/proc/intensity_update()
+	if(effect_power >= 50) //If we're at the threshold for the highest tier effect, we change sprites in preparation for the spooks.
+		icon_state = "ectoplasm_heavy"
+		update_icon_state()
+	else
+		icon_state = "ectoplasm"
+		update_icon_state()
+
+
+
+// Ghost Portal. Used to bring anomaly orbiters into the playing field as ghosts. Destroys itself and all of its associated ghosts after two minutes.
+// Can be destroyed early to the same effect.
+
+/obj/structure/ghost_portal
+	name = "Spooky Portal"
+	desc = "A portal between our dimension and who-knows-where? It's emitting an absolutely ungodly wailing sound."
+	icon = 'icons/obj/objects.dmi'
+	icon_state = "anom"
+	anchored = TRUE
+	var/static/list/spooky_noises = list(
+		'sound/hallucinations/growl1.ogg',
+		'sound/hallucinations/growl2.ogg',
+		'sound/hallucinations/growl3.ogg',
+		'sound/hallucinations/veryfar_noise.ogg',
+		'sound/hallucinations/wail.ogg'
+	)
+	var/list/ghosts_spawned = list()
+
+/obj/structure/ghost_portal/Initialize(mapload, candidate_list)
+	. = ..()
+
+	START_PROCESSING(SSobj, src)
+	INVOKE_ASYNC(src, PROC_REF(make_ghost_swarm), candidate_list)
+	playsound(src, pick(spooky_noises), 100, TRUE)
+	QDEL_IN(src, 2 MINUTES)
+
+/obj/structure/ghost_portal/process(delta_time)
+	. = ..()
+
+	if(prob(5))
+		playsound(src, pick(spooky_noises), 100)
+
+/obj/structure/ghost_portal/play_attack_sound(damage_amount, damage_type = BRUTE, damage_flag = 0)
+	playsound(loc, 'sound/effects/empulse.ogg', 75, TRUE)
+	if(prob(40))
+		playsound(src, pick(spooky_noises), 50)
+
+/obj/structure/ghost_portal/Destroy()
+	. = ..()
+
+	STOP_PROCESSING(SSobj, src)
+	cleanup_ghosts()
```

Comment:
I'm not a fan of this proc being used in Destroy() because it does a bit more than just clean up our references (e.g. "fluffy" stuff) like playing sounds, spawning an effect...

I would say have this call `cleanup_ghosts()` asynchronously (maybe with a one tick wait or something?) with the `ghosts_spawned` list of all of the ghosts it has within it, _then_ set the `ghosts_spawned` variable to null/empty list(). You would have to update `cleanup_ghosts()` to accept a list of ghosts as an argument (as well as move the proc to a global (or send out a signal that the swarm ghosts will listen to and do it themselves) since our src will be deleted), but I think it's much better to do it this way to ensure we don't clog up the garbage collector with stuff we don't need.

i could be full of shit but its just this Destroy is doing more heavy lifting than it should be reasonably expected to. If you know of a better way to handle this, do that, but just try to not give so much to chew on when we're trying to get our shit cleaned up